Firestore/core/test/unit/local/leveldb_local_store_test.cc

```diff
@@ -322,6 +324,435 @@ TEST_F(LevelDbLocalStoreTest, IndexesServerTimestamps) {
+      testutil::Query("coll").AddingFilter(Filter("value", "==", "match"));
+  int target_id = AllocateQuery(query);
+
+  SetIndexAutoCreationEnabled(true);
+  SetMinCollectionSizeToAutoCreateIndex(0);
+  SetRelativeIndexReadCostPerDocument(2);
+
+  ApplyRemoteEvent(
+      AddedRemoteEvent(Doc("coll/a", 10, Map("value", "match")), {target_id}));
+  ApplyRemoteEvent(AddedRemoteEvent(
+      Doc("coll/b", 10, Map("value", std::numeric_limits<double>::quiet_NaN())),
+      {target_id}));
+  ApplyRemoteEvent(
+      AddedRemoteEvent(Doc("coll/c", 10, Map("value", nullptr)), {target_id}));
+  ApplyRemoteEvent(AddedRemoteEvent(Doc("coll/d", 10, Map("value", "mismatch")),
+                                    {target_id}));
+  ApplyRemoteEvent(
+      AddedRemoteEvent(Doc("coll/e", 10, Map("value", "match")), {target_id}));
+
+  // First time query is running without indexes.
+  // Based on current heuristic, collection document counts (5) > 2 * resultSize
+  // (2). Full matched index should be created.
+  ExecuteQuery(query);
+  FSTAssertRemoteDocumentsRead(/* byKey= */ 0, /* byCollection= */ 2);
+  FSTAssertQueryReturned("coll/a", "coll/e");
+
+  BackfillIndexes();
+
+  ExecuteQuery(query);
+  FSTAssertRemoteDocumentsRead(/* byKey= */ 2, /* byCollection= */ 0);
+  FSTAssertQueryReturned("coll/a", "coll/e");
+
+  DeleteAllIndexes();
+
+  ExecuteQuery(query);
+  FSTAssertRemoteDocumentsRead(/* byKey= */ 0, /* byCollection= */ 2);
+  FSTAssertQueryReturned("coll/a", "coll/e");
+
+  // Field index is created again.
+  BackfillIndexes();
+
+  ExecuteQuery(query);
+  FSTAssertRemoteDocumentsRead(/* byKey= */ 2, /* byCollection= */ 0);
+  FSTAssertQueryReturned("coll/a", "coll/e");
+}
+
+TEST_F(LevelDbLocalStoreTest, DeleteAllIndexesWorksWithManualAddedIndexes) {
+  FieldIndex index =
+      MakeFieldIndex("coll", 0, FieldIndex::InitialState(), "matches",
+                     model::Segment::Kind::kAscending);
+  ConfigureFieldIndexes({index});
+
+  core::Query query =
+      testutil::Query("coll").AddingFilter(Filter("matches", "==", true));
+  int target_id = AllocateQuery(query);
+
+  ApplyRemoteEvent(
+      AddedRemoteEvent(Doc("coll/a", 10, Map("matches", true)), {target_id}));
+
+  BackfillIndexes();
+
+  ExecuteQuery(query);
+  FSTAssertRemoteDocumentsRead(/* byKey= */ 1, /* byCollection= */ 0);
+  FSTAssertQueryReturned("coll/a");
+
+  DeleteAllIndexes();
+
+  ExecuteQuery(query);
+  FSTAssertRemoteDocumentsRead(/* byKey= */ 0, /* byCollection= */ 1);
+  FSTAssertQueryReturned("coll/a");
+}
+
+TEST_F(LevelDbLocalStoreTest, DeleteAllIndexesWorksWith1000Documents) {
+  FieldIndex index =
+      MakeFieldIndex("coll", 0, FieldIndex::InitialState(), "matches",
+                     model::Segment::Kind::kAscending);
+  ConfigureFieldIndexes({index});
+
+  core::Query query =
+      testutil::Query("coll").AddingFilter(Filter("matches", "==", true));
+  int target_id = AllocateQuery(query);
+
+  for (int count = 1; count <= 1000; count++) {
```

Comment:
Should we increase this to a number above 1000, like 1100 to make sure second transaction also runs fine?

Reply:
Also is it possible to use the actual constant instead of a magic number?